small/ibuf.c

```diff
@@ -80,27 +99,29 @@ ibuf_reserve_slow(struct ibuf *ibuf, size_t size)
 	 */
 	if (size + used <= capacity) {
 		memmove(ibuf->buf, ibuf->rpos, used);
+		ibuf->rpos = ibuf->buf;
+		ibuf->wpos = ibuf->rpos + used;
+		return ibuf->wpos;
 	} else {
-		/* Use iobuf_readahead as allocation factor. */
+		/* Use start_capacity as allocation factor. */
 		size_t new_capacity = capacity * 2;
 		if (new_capacity < ibuf->start_capacity)
 			new_capacity = ibuf->start_capacity;
 
 		while (new_capacity < used + size)
 			new_capacity *= 2;
 
-		struct slab *slab = slab_get(ibuf->slabc, new_capacity);
-		if (slab == NULL)
-			return NULL;
-		char *ptr = (char *) slab_data(slab);
-		memcpy(ptr, ibuf->rpos, used);
-		if (ibuf->buf)
-			slab_put(ibuf->slabc, slab_from_data(ibuf->buf));
-		ibuf->buf = ptr;
-		ibuf->end = ibuf->buf + slab_capacity(slab);
+		return ibuf_resize(ibuf, new_capacity);
 	}
-	ibuf->rpos = ibuf->buf;
-	ibuf->wpos = ibuf->rpos + used;
-	return ibuf->wpos;
 }
 
+void
+ibuf_shrink(struct ibuf *ibuf)
+{
+	size_t new_capacity = ibuf_used(ibuf);
+	if (new_capacity < ibuf->start_capacity)
+		new_capacity = ibuf->start_capacity;
+
+	/* On memory allocation failure, the buffer is not modified. */
+	ibuf_resize(ibuf, new_capacity);
```

Comment:
I don't think we should do anything if the new capacity equals the old one.

Reply:
Done